source/lac/petsc_solver.cc

```diff
@@ -709,154 +683,121 @@ namespace PETScWrappers
                            const VectorBase &b)
   {
 #  ifdef DEAL_II_PETSC_WITH_MUMPS
-    /*
-     * factorization matrix to be obtained from MUMPS
-     */
-    Mat F;
-
-    /*
-     * setting MUMPS integer control parameters ICNTL to be passed to
-     * MUMPS.  Setting entry 7 of MUMPS ICNTL array (of size 40) to a value
-     * of 2. This sets use of Approximate Minimum Fill (AMF)
-     */
-    PetscInt ival = 2, icntl = 7;
-    /*
-     * number of iterations to solution (should be 1) for a direct solver
-     */
-    PetscInt its;
-    /*
-     * norm of residual
-     */
-    PetscReal rnorm;
-
     /*
      * creating a solver object if this is necessary
      */
-    if (solver_data == nullptr)
+    if (ksp == nullptr)
       {
-        solver_data = std::make_unique<SolverDataMUMPS>();
-
         /*
-         * creates the default KSP context and puts it in the location
-         * solver_data->ksp
+         * factorization matrix to be obtained from MUMPS
          */
-        PetscErrorCode ierr =
-          KSPCreate(A.get_mpi_communicator(), &solver_data->ksp);
-        AssertThrow(ierr == 0, ExcPETScError(ierr));
+        Mat F;
 
         /*
-         * set the matrices involved. the last argument is irrelevant here,
-         * since we use the solver only once anyway
+         * setting MUMPS integer control parameters ICNTL to be passed to
+         * MUMPS.  Setting entry 7 of MUMPS ICNTL array (of size 40) to a value
+         * of 2. This sets use of Approximate Minimum Fill (AMF)
          */
-        ierr = KSPSetOperators(solver_data->ksp, A, A);
-        AssertThrow(ierr == 0, ExcPETScError(ierr));
+        PetscInt ival = 2, icntl = 7;
```

Comment:
This might be easier to read if you move the declaration of `ival` and `icntl` down right above the place where you use these variables. Right now definition and use are separated by quite a lot of vertical space (=mental space).

Reply:
those changes are hard to read because I also removed useless variables. I'll make it as clear as possible